### ISSUE TITLE
Acceptance criteria

### ISSUE THREAD

**Issue by muratcorlu** [muratcorlu/awesome-forms#1]
We need to define an "acceptance criteria" for form examples. Is this a good use case?

---

1. You will implement a complex registration form.
1. Fields:
    1. Name. Required. At least 2 words. Maximum 100 chars. Numbers not allowed.
    1. Email. Required. Email validation. Server-side request[*](#e-mail-check-endpoint)(when input changed) for checking if it's already registered.
    1. Country Selection. Required. Select from a list with `id: label`. `id` value should sent to server
    1. City Selection. Optional. Select from a list with `id: label`. `id` value should sent to server. City selection should be shown after `country` selection.
    1. Photo. Optional. Image upload. Only jpg format should be allowed. File size should be less than 5MB.
    1. Password. 2 fields should be matched. At least 8 chars that includes one number, one uppercase letter, one lowercase letter.
1. Form output to server should be json in that format:
    ```json
    {
      "name": "Yan Chu",
      "email": "awesome@email.com",
      "country": "90",
      "city": "34",
      "photo": "http://to.an/image/url.jpg",
      "password": "Crazy-82"
    }
    ```
1. Form should be prefillable with same output json object.

### E-mail check endpoint

GET: https://somedomain.now.sh/emails/{email@address.com}

Will return `404` if this email is not registered. (Without response body)
Will return `200` if this email is registered. (Without response body)

**Comment by ismail-codar** [comment]
Address entry with two different ways.
 1- City, district, neighborhood selection and door number entry. (Address textfield is disabled and value of address is calculated from selectboxes)
 2- Free-text address entry (Address textfield is enabled and City, district, neighborhood, door number fields are hidden)

**Comment by ismail-codar** [comment]
If "Is Active" is checked, the "Password Expiration Date" field will be displayed and the entry will be required. Otherwise, the "Password Expiration Date" field must be hidden.

**Comment by ismail-codar** [comment]
Form UI (html and css) must be pre defined in advance. All implementations should use it.
UI Features:
1- Mandatory fields must be indicated by (*)
2- In the wrong entries, the .error css-class is added on the field. (with live validation)
3- Inline error message is putted below the field.

**Comment by muratcorlu** [comment]
> Address entry with two different ways.
> 1- City, district, neighborhood selection and door number entry. (Address textfield is disabled and value of address is calculated from selectboxes)
> 2- Free-text address entry (Address textfield is enabled and City, district, neighborhood, door number fields are hidden)

City, district, neighborhood selection needs a bit more complicated backend support. Also doesn't add a specific feature for comparison. Couldn't be sure about that.

**Comment by muratcorlu** [comment]
> If "Is Active" is checked, the "Password Expiration Date" field will be displayed and the entry will be required. Otherwise, the "Password Expiration Date" field must be hidden.

Having a date/time field can be a good idea to compare date validation approaches.

**Comment by muratcorlu** [comment]
> Form UI (html and css) must be pre defined in advance. All implementations should use it.
> UI Features:
> 1- Mandatory fields must be indicated by (*)
> 2- In the wrong entries, the .error css-class is added on the field. (with live validation)
> 3- Inline error message is putted below the field.

Giving a starter HTML/CSS page could be helpful for excluding UI design part from comparison. But forcing people to use this HTML/CSS only doesn't look like a good idea. Maybe some people will want to use custom components with different HTML structure.

**Comment by ismail-codar** [comment]
> > Address entry with two different ways.
> > 1- City, district, neighborhood selection and door number entry. (Address textfield is disabled and value of address is calculated from selectboxes)
> > 2- Free-text address entry (Address textfield is enabled and City, district, neighborhood, door number fields are hidden)
> 
> City, district, neighborhood selection needs a bit more complicated backend support. Also doesn't add a specific feature for comparison. Couldn't be sure about that.


Browser side mock [server](https://github.com/pretenderjs/pretender) can be a solution for backend api.

My intention is form logic should not be simple and can have these features. Needed unnecessary ones can be discussed
 - Programmatically loading data into select lists (combo, radibuttons etc). A cascading data [example](https://www.cognitoforms.com/blog/176/cascading-lookups)
 - Conditionally show hide or enable disable fields if required. ([example](https://www.formstack.com/features/conditional-logic))
 - Live validation with oninput, onblur events or validation on form post.
 - Form data edit mode, reset form to default values, readonly mode.
 - Hidden fields.
 - File or file list upload. Also inline display for selected file.
 - Sub form or input lists.
 -  ...

**Comment by burakcan** [comment]
I think the examples shouldn't be using any third party libraries besides the fw itself. This way it'd be easier to see and compare differences.

**Comment by muratcorlu** [comment]
AFAIK, Vue and React doesn't give a special solution for form validation. But many developers of them prefers to use some plugins. I think it would be good to show all options, including core library solutions.

**Comment by burakcan** [comment]
So maybe having them under seperate entries could be an option. Like: "React, React-xlib, React-ylib"

**Comment by muratcorlu** [comment]
Indeed. My idea was having separate examples for different approaches or extensions.

**Comment by ismail-codar** [comment]
Angular ngx-formly, react-final-form, formik or vue-final-form... Third party components can be used for form validation, form state management etc..

My opinion is that the frameworks-libs-infrastructure can be different but the form behavior and user interface should be the same for all implementations. It like [TodoMVC](http://todomvc.com/). 
My suggestion: Pre-define form UI with [bulma](https://bulma.io/documentation/form/general/) or with bootstrap form controls.  Also  standardize form behavior. 
Only third party UI libraries should not be used.

Also another idea is a standardized form behavior but user interface can be freely different.

**Comment by muratcorlu** [comment]
Then let's create a form with just html and css as a reference. Then everyone can do their own versions. I'll try to create that reference form.
